File: src/lib/data/board.ts
import { join } from 'node:path';

import type { LoadedDataset, LoadIssue } from './load.ts';
import type { ResultAsync } from 'neverthrow';

import { type BoardCategory, BoardFile } from '../schemas/board.ts';
import { parse } from './util.ts';
import { queryPool } from './query.ts';

/** A resolved 2D grid — `grid[categoryIndex][pointIndex] = questionId | null`. */
export type BoardGrid = (string | null)[][];

export interface BuildBoardOptions {
	locale: string;
	seed?: number;
}

/** Load and validate a board YAML file. Returns err with issues on failure. */
export function loadBoard(dir: string, filename: string): ResultAsync<BoardFile, LoadIssue[]> {
	return parse(join(dir, filename), BoardFile);
}

/**
 * Build a resolved NxM board. Returns `grid[ci][pi] = questionId | null`.
 * Unresolvable slots are `null` and reported as issues.
 */
export function buildBoard(
	ds: LoadedDataset,
	board: BoardFile,
	opts: BuildBoardOptions
): { grid: BoardGrid; issues: LoadIssue[] } {
	const issues: LoadIssue[] = [];
	const grid: BoardGrid = [];
	const rng = mulberry32(opts.seed ?? 42);
	const diffMap = board.difficulty_map ?? {};

	const packCache = new Map<string, string[]>();

	const used = new Set<string>();

	for (const cat of board.categories) {
		const row: (null | string)[] = [];
		for (const point of board.points.entries()) {
			const pointKey = String(point);

			const explicit = cat.question_ids?.[pointKey];
			if (explicit) {
				row.push(explicit);
				used.add(explicit);
				continue;
			}

			const candidates = buildCandidates(cat, pointKey, ds, packCache, diffMap);
			const unused = candidates.filter((id) => !used.has(id));
			const pool = unused.length > 0 ? unused : candidates;

			const [picked = null] = pool.length > 0 ? shuffle(pool, rng) : [];
			row.push(picked);
			if (picked) used.add(picked);
		}
		grid.push(row);
	}

	return { grid, issues };
}

function shuffle<T>(arr: T[], rng: () => number): T[] {
	const out = [...arr];
	for (let i = out.length - 1; i > 0; i--) {
		const j = Math.floor(rng() * (i + 1));
		const temp = out[i];
		if (out[j] != null) out[i] = out[j];
		if (temp != null) out[j] = temp;
	}
	return out;
}

/** Deterministic seeded PRNG (mulberry32). */
function mulberry32(a: number) {
	return () => {
		a |= 0;
		a = (a + 0x6d2b79f5) | 0;
		let t = Math.imul(a ^ (a >>> 15), 1 | a);
		t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
		return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
	};
}

function buildCandidates(
	cat: BoardCategory,
	pointKey: string,
	ds: LoadedDataset,
	packCache: Map<string, string[]>,
	diffMap: Record<string, string[] | undefined>
): string[] {
	let candidates: string[] = [];

	if (cat.pack_ref) {
		candidates = resolvePack(ds, packCache, cat.pack_ref);
	}

	if (cat.filter) {
		const filteredSet = new Set(queryPool(ds, cat.filter));
		candidates =
			candidates.length > 0 ? candidates.filter((qid) => filteredSet.has(qid)) : [...filteredSet];
	}

	const diffTags = diffMap[pointKey];
	if (diffTags && diffTags.length > 0) {
		candidates = candidates.filter((qid) => {
			const tags = ds.questions.get(qid)?.item.tags;
			return tags !== undefined && diffTags.some((tag) => tags.includes(tag));
		});
	}

	return candidates;
}

function resolvePack(
	ds: LoadedDataset,
	packCache: Map<string, string[]>,
	packId: string
): string[] {
	const cached = packCache.get(packId);
	if (cached !== undefined) return cached;

	const ids: string[] = [];
	const pack = ds.packs.get(packId);
	if (!pack) return ids;
	// includes
	for (const incl of pack.item.includes ?? []) {
		ids.push(...resolvePack(ds, packCache, incl));
	}
	// explicit questions
	ids.push(...(pack.item.questions ?? []));
	// pack filter query
	if (pack.item.filter) {
		ids.push(...queryPool(ds, pack.item.filter));
	}
	const unique = [...new Set(ids)];
	packCache.set(packId, unique);
	return unique;
}

File: api/src/data/board.rs
//! Board builder — resolve a board definition into a 2D grid of question IDs.
//!
//! Mirrors `board.ts`: explicit IDs win, then pack refs, then filters.
//! Deterministic shuffle via mulberry32 PRNG.

use std::collections::{HashMap, HashSet};

use super::query::query_pool;
use super::types::*;

/// Resolved board: `grid[category_idx][point_idx] = Some(question_id) | None`.
pub type BoardGrid = Vec<Vec<Option<String>>>;

pub struct BuildBoardOpts {
    pub locale: String,
    pub seed: u32,
}

impl Default for BuildBoardOpts {
    fn default() -> Self {
        Self {
            locale: "en".into(),
            seed: 42,
        }
    }
}

/// Build a resolved NxM board grid. Unresolvable slots are `None`.
pub fn build_board(
    ds: &LoadedDataset,
    board: &BoardFile,
    opts: &BuildBoardOpts,
) -> (BoardGrid, Vec<LoadIssue>) {
    let issues = Vec::new();
    let mut rng = Mulberry32::new(opts.seed);
    let mut used = HashSet::new();
    let mut pack_cache: HashMap<String, Vec<String>> = HashMap::new();

    let diff_map: HashMap<&str, &[String]> = board
        .difficulty_map
        .as_ref()
        .map(|dm| {
            dm.iter()
                .map(|(k, v)| (k.as_str(), v.as_slice()))
                .collect()
        })
        .unwrap_or_default();

    let mut grid = Vec::new();

    for cat in &board.categories {
        let mut row = Vec::new();
        for point in &board.points {
            let point_key = point.to_string();

            // 1. Explicit question_ids override
            if let Some(qid) = cat.question_ids.as_ref().and_then(|m| m.get(&point_key)) {
                used.insert(qid.clone());
                row.push(Some(qid.clone()));
                continue;
            }

            // 2. Build candidates from pack_ref + filter + difficulty_map
            let candidates = build_candidates(cat, &point_key, ds, &mut pack_cache, &diff_map);
            let unused: Vec<&String> = candidates.iter().filter(|id| !used.contains(*id)).collect();
            let pool: Vec<&String> = if unused.is_empty() {
                candidates.iter().collect()
            } else {
                unused
            };

            if let Some(picked) = pick_random(&pool, &mut rng) {
                used.insert(picked.clone());
                row.push(Some(picked));
            } else {
                row.push(None);
            }
        }
        grid.push(row);
    }

    (grid, issues)
}

fn build_candidates(
    cat: &BoardCategory,
    point_key: &str,
    ds: &LoadedDataset,
    pack_cache: &mut HashMap<String, Vec<String>>,
    diff_map: &HashMap<&str, &[String]>,
) -> Vec<String> {
    let mut candidates = Vec::new();

    if let Some(ref pack_id) = cat.pack_ref {
        candidates = resolve_pack(ds, pack_cache, pack_id);
    }

    if let Some(ref filter) = cat.filter {
        let filtered: HashSet<String> = query_pool(ds, filter).into_iter().collect();
        if candidates.is_empty() {
            candidates = filtered.into_iter().collect();
        } else {
            candidates.retain(|id| filtered.contains(id));
        }
    }

    if let Some(diff_tags) = diff_map.get(point_key) {
        if !diff_tags.is_empty() {
            candidates.retain(|qid| {
                ds.questions
                    .get(qid)
                    .map(|e| diff_tags.iter().any(|tag| e.item.tags().contains(&tag.as_str().to_owned())))
                    .unwrap_or(false)
            });
        }
    }

    candidates
}

fn resolve_pack(
    ds: &LoadedDataset,
    cache: &mut HashMap<String, Vec<String>>,
    pack_id: &str,
) -> Vec<String> {
    if let Some(cached) = cache.get(pack_id) {
        return cached.clone();
    }

    let Some(entry) = ds.packs.get(pack_id) else {
        return Vec::new();
    };
    let pack = &entry.item;

    let mut ids = Vec::new();

    for incl in pack.includes.iter().flatten() {
        ids.extend(resolve_pack(ds, cache, incl));
    }
    ids.extend(pack.questions.iter().flatten().cloned());
    if let Some(ref filter) = pack.filter {
        ids.extend(query_pool(ds, filter));
    }

    // Deduplicate preserving order
    let mut seen = HashSet::new();
    ids.retain(|id| seen.insert(id.clone()));

    cache.insert(pack_id.to_owned(), ids.clone());
    ids
}

fn pick_random(pool: &[&String], rng: &mut Mulberry32) -> Option<String> {
    if pool.is_empty() {
        return None;
    }
    let mut shuffled: Vec<&String> = pool.to_vec();
    for i in (1..shuffled.len()).rev() {
        let j = (rng.next() * (i + 1) as f64) as usize;
        shuffled.swap(i, j);
    }
    shuffled.first().map(|s| (*s).clone())
}

// ─── Mulberry32 PRNG ────────────────────────────────────────────────────────

struct Mulberry32 {
    state: u32,
}

impl Mulberry32 {
    fn new(seed: u32) -> Self {
        Self { state: seed }
    }

    fn next(&mut self) -> f64 {
        self.state = self.state.wrapping_add(0x6D2B_79F5);
        let mut t = self.state;
        t = (t ^ (t >> 15)).wrapping_mul(1 | t);
        t = (t.wrapping_add((t ^ (t >> 7)).wrapping_mul(61 | t))) ^ t;
        ((t ^ (t >> 14)) as f64) / 4_294_967_296.0
    }
}
